qualang_tools/config/server/upload.py

```diff
@@ -0,0 +1,165 @@
+        dcc.Upload(
+            id="upload-config-edits",
+            children=html.Div(
+                [
+                    "Optionally, any previously generated ",
+                    html.Strong("config_edits.py"),
+                    "(Drag and Drop or Click to select)",
+                ]
+            ),
+            style={
+                "width": "100%",
+                "height": "60px",
+                "lineHeight": "60px",
+                "borderWidth": "1px",
+                "borderStyle": "dashed",
+                "borderRadius": "5px",
+                "textAlign": "center",
+                "margin": "10px",
+            },
+            # Allow multiple files to be uploaded
+            multiple=False,
+        ),
+        html.Div(id="output-data-edits-upload"),
+    ]
+)
+
+
+@app.callback(
+    [Output("output-data-upload", "children")],
+    Input("upload-config", "contents"),
+    State("upload-config", "filename"),
+    State("upload-config", "last_modified"),
+)
+def upload_init_config(contents, filename, last_modified):
+    if contents is not None:
+        init_edits_file()
+        save_file(os.path.join(UPLOAD_DIRECTORY, "config_initial.py"), contents)
+        print("OK")
+        return [html.P(f"Uploaded {filename}")]
+    return [""]
+
+
+@app.callback(
+    [Output("output-data-edits-upload", "children")],
+    Input("upload-config-edits", "contents"),
+    State("upload-config-edits", "filename"),
+    State("upload-config-edits", "last_modified"),
+)
+def upload_config_edits(contents, filename, last_modified):
+    if contents is not None:
+        save_file(os.path.join(UPLOAD_DIRECTORY, "config_edits.py"), contents)
+        return [html.P(f"Uploaded {filename}")]
+    return [""]
+
+
+def save_file(name, content):
+    """Decode and store a file uploaded with Plotly Dash."""
+    data = content.encode("utf8").split(b";base64,")[1]
+    with open(os.path.join(UPLOAD_DIRECTORY, name), "wb") as fp:
+        fp.write(base64.decodebytes(data))
+
+
+@app.callback(
+    Output("config-initial-status", "children"),
+    Input("button-empty-config", "n_clicks"),
+    prevent_initial_call=True,
+)
+def enter_empty_config(click):
+    with open(os.path.join(UPLOAD_DIRECTORY, "config_initial.py"), "w") as fp:
+        fp.write(
+            """
+configuration = {
+    "version": 1,
+    "controllers": {},
+    "elements": {},
+    "pulses": {},
+    "waveforms": {},
+    "digital_waveforms": {},
+    "integration_weights": {},
+    "oscillators": {},
+    "mixers": {},
+}
+"""
+        )
+    init_edits_file()
+
+    with open(os.path.join(UPLOAD_DIRECTORY, "config_final.py"), "w") as fp:
+        fp.write(
+            """
+from config_edits import configuration, setup
+
+configuration = setup.build(configuration)
+"""
+        )
+    return "Starting from empty configuration..."
+
+
+def init_edits_file():
+    with open(os.path.join(UPLOAD_DIRECTORY, "config_edits.py"), "w") as fp:
+        fp.write(
```

Comment:
can you add tests that validate files are created correctly and can be loaded to python?
(in the tests folder, maybe create a function that do it and call it from the callback.

Reply:
Added in 9ed79ba98943b27a46d7dfa55fbdff4f0470cd03